src/QMCWaveFunctions/detail/CUDA/cuBLAS_LU.cu

```diff
@@ -0,0 +1,212 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2021 QMCPACK developers.
+//
+// File developed by: Peter Doak, doakpw@ornl.gov, Oak Ridge National Laboratory
+//
+// File created by: Peter Doak, doakpw@ornl.gov, Oak Ridge National Laboratory
+//////////////////////////////////////////////////////////////////////////////////////
+
+#include "cuBLAS_LU.hpp"
+#include "Platforms/CUDA/cudaError.h"
+#include "Platforms/CUDA/cuBLAS.hpp"
+
+namespace qmcplusplus
+{
+namespace cuBLAS_LU
+{
+
+__device__ cuDoubleComplex complexDetLog(const double lu_diag, const int n_index, const int* pivots)
+{
+  cuDoubleComplex log_value;
+  log_value.x = log(abs(lu_diag));
+  log_value.y = ((lu_diag < 0) != ((pivots[n_index] - 1) == n_index)) * M_PI;
+  return log_value;
+}
+
+__device__ cuDoubleComplex complexDetLog(const cuDoubleComplex lu_diag, const int n_index, const int* pivots)
+{
+  cuDoubleComplex diag;
+  double pivot_factor = 1 - 2 * ((pivots[n_index] - 1) == n_index);
+  diag.x              = lu_diag.x * pivot_factor;
+  diag.y              = lu_diag.y * pivot_factor;
+  cuDoubleComplex log_value;
+  log_value.x = log(sqrt(diag.x * diag.x + diag.y * diag.y));
+  log_value.y = atan2(diag.y, diag.x);
+  return log_value;
+}
+
+/** computes the log determinant using the output of LU factorization.
+   */
+template<int COLBS, typename T>
+__global__ void computeLogDet_kernel(const int n,
+                                     const int lda,
+                                     T** mat_lus,
+                                     const int* const pivots,
+                                     cuDoubleComplex* logdets)
+{
+  const int iw                = blockIdx.x;
+  const int block_num         = blockIdx.y;
+  const T* __restrict__ lu_iw = mat_lus[iw];
+  __shared__ cuDoubleComplex logdet_vals[COLBS];
+  T lu_diag;
+  int n_index = threadIdx.x + block_num * COLBS;
+  if (n_index < n)
+  {
+    lu_diag                  = *(lu_iw + n_index * lda + n_index);
+    logdet_vals[threadIdx.x] = complexDetLog(lu_diag, n_index, pivots);
+  }
+  if (threadIdx.x == 0 && blockIdx.y == 0)
+    logdets[iw] = {0.0, 0.0};
+  // insure that when we reduce logdet_vals all the threads in the block are done.
+  __syncthreads();
+  if (threadIdx.x == 0)
```

Comment:
spawn one block per COLBS is not necessary, this kernel anyway doesn't have much computation.
I expect one walker per block is good enough. By doing that, we can avoid atomic operations since everything is within a block. The n iterations is still done by loop over COLBS segments and then do a binary tree reduction.

Reply:
This can be explored once I have profiling set up, for now I think this is the best way to implement this and there is no good reason to rewrite it while the rest of the code already developed for this feature diverges.

Reply:
I mean
`dimGrid(batch_size)` instead of `dimGrid(batch_size, num_col_blocks)`
So no more need of atomics. one full matrix per block.
Yes. We can look at this later.